### ISSUE TITLE
Client/server enhancement ideas

### ISSUE THREAD

**Issue by hawkrives** [hybsearch/hybsearch#98]
- Maintain list of ongoing/completed runs
- Show list of ongoing/completed runs in the client
- Support multiple clients watching the same run

---

- Stream terminal output to clients like CircleCI does for more immediate feedback
- Allow clients to view outputs of each step of the pipeline - maybe by tapping on the progress icon?
- Implement more of a graph-like view of the progress indicators, instead of the linear list, to show interconnections

---

I think I had another. Maybe it'll come to me.

**Comment by hawkrives** [comment]
Agreed on the graph view. I'll split this apart into several issues later today, so we can prioritize appropriately.

**Comment by rye** [comment]
>- Maintain list of ongoing/completed runs
>- Show list of ongoing/completed runs in the client
>- Support multiple clients watching the same run

I think that generalizing this feature to more broadly mean "track run state" would mean that getting a list of current/past runs is simply an index operation in a REST-y sense.  I think we should implement this like this:

- When the client indexes current job runs, the client connects to `wss?://<server-uri>/api/runs`.  On connection, server sends back current state to the client.  WS connection is then added to a pub/sub pool; whenever a job state changes or a job completes, the client gets a new message with either what changed, or the new state.

- Job statuses can be done similarly.  Client would connect to something à la `wss?://<server-uri>/api/run/<random data>`, and they would get fed: (i) the current state of the server, and (ii) subsequent updates.

To do the above, the server just needs to keep track of state, and then attach state change handlers so that it can send messages to all clients.

>- Stream terminal output to clients like CircleCI does for more immediate feedback

I like this feature, as it would slot in nicely with the above three features.  ("Updates to state" could include new console output, for instance.)

>- Allow clients to view outputs of each step of the pipeline - maybe by tapping on the progress icon?

I think we can extend the API idea that I wrote on upstairs to include this; we could expose REST/GraphQL endpoints for this.

>- Implement more of a graph-like view of the progress indicators, instead of the linear list, to show interconnections

I agree with @OmarShehata—this would be a nice longer-term thing, but definitely should not be a blocker.
